plugins/fpc/fu-fpc-device.c

```diff
@@ -0,0 +1,583 @@
+/*
+ * Copyright (C) 2022 Haowei Lo <haowei.lo@fingerprints.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-fpc-device.h"
+
+#define FPC_USB_INTERFACE	     0
+#define FPC_USB_TRANSFER_TIMEOUT     1500 /* ms */
+#define FPC_FLASH_BLOCK_SIZE_DEFAULT 2048 /* 2048 */
+#define FPC_FLASH_BLOCK_SIZE_4096    4096 /* 4096 */
+
+#define FPC_CMD_DFU_DETACH	  0x00
+#define FPC_CMD_DFU_DNLOAD	  0x01
+#define FPC_CMD_DFU_GETSTATUS	  0x03
+#define FPC_CMD_DFU_CLRSTATUS	  0x04
+#define FPC_CMD_DFU_GET_FW_STATUS 0x09
+
+#define FPC_CMD_BOOT0	  0x04
+#define FPC_CMD_GET_STATE 0x0B
+
+#define FPC_DEVICE_MOC_STATE_LEN     68
+#define FPC_DEVICE_MOH_STATE_LEN     72
+#define FPC_DEVICE_DFU_FW_STATUS_LEN 8
+#define FPC_DFU_MAX_ATTEMPTS	     50
+
+#define FPC_DEVICE_DFU_MODE_CLASS    0xFE
+#define FPC_DEVICE_DFU_MODE_PORT     0x02
+#define FPC_DEVICE_NORMAL_MODE_CLASS 0xFF
+#define FPC_DEVICE_NORMAL_MODE_PORT  0xFF
+
+#define FPC_DFU_DNBUSY 0x04
+
+/**
+ * FU_FPC_DEVICE_FLAG_MOH_DEVICE:
+ *
+ * Device is a moh device
+ */
+#define FU_FPC_DEVICE_FLAG_MOH_DEVICE (1 << 0)
+
+/**
+ * FU_FPC_DEVICE_FLAG_LEGACY_DFU:
+ *
+ * Device supports legacy dfu mode
+ */
+
+#define FU_FPC_DEVICE_FLAG_LEGACY_DFU (1 << 1)
+
+/**
+ * FU_FPC_DEVICE_FLAG_RTS_DEVICE:
+ *
+ * Device is a RTS device
+ */
+
+#define FU_FPC_DEVICE_FLAG_RTS_DEVICE (1 << 2)
+
+struct _FuFpcDevice {
+	FuUsbDevice parent_instance;
+	guint32 max_block_size;
+};
+
+G_DEFINE_TYPE(FuFpcDevice, fu_fpc_device, FU_TYPE_USB_DEVICE)
+
+typedef struct __attribute__((packed)) {
+	guint8 bstatus;
+	guint8 bmax_payload_size;
+	guint8 reserved[2];
+	guint8 bstate;
+	guint8 reserved2;
+} FuFpcDfuStatus;
+
+static gboolean
+fu_fpc_device_dfu_cmd(FuFpcDevice *self,
+		      guint8 request,
+		      guint16 value,
+		      guint8 *data,
+		      gsize length,
+		      gboolean device2host,
+		      gboolean type_vendor,
+		      GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	gsize actual_len = 0;
+	gboolean ret;
+
+	if (data == NULL && length > 0) {
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Invalid input data");
+		return FALSE;
+	}
+
+	ret = g_usb_device_control_transfer(usb_device,
+					    device2host ? G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST
+							: G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    type_vendor ? G_USB_DEVICE_REQUEST_TYPE_VENDOR
+							: G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					    G_USB_DEVICE_RECIPIENT_INTERFACE,
+					    request,
+					    value,
+					    0x0000,
+					    data,
+					    length,
+					    length ? &actual_len : NULL,
+					    FPC_USB_TRANSFER_TIMEOUT,
+					    NULL,
+					    error);
+	if (!ret)
+		return FALSE;
```

Comment:
remove the temp `ret` variable please, just do a big `if (!g_usb_device_control_transfer`